### DISCUSSION TITLE
post/durable-functions-multiple-external-events

### DISCUSSION THREAD

**Discussion by giscus** [markheath/soundcode-blog-comments#21] [Announcements]
# post/durable-functions-multiple-external-events

Mark Heath's Development Blog

https://markheath.net/post/durable-functions-multiple-external-events

**Comment by andsj073** [comment]
Hey Mark! Excellent work and articles on Durable FX and external events!
I understand that ext. events are queued in-memory in Durable FX. 
Is that a risk here as the container and/or host could (and will be) be restarted at some point in its life-cycle?
Are there any mitigations to that to guarantee event delivery?

**Reply by markheath** [reply]
I don't think external events are queued in-memory. Was there somewhere that you read that? Durable Functions uses Azure Storage by default with Table Storage holding the event-sourced history of an orchestration, and Azure Storage Queues for triggering the underlying functions.

**Reply by andsj073** [reply]
I read it here: https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-external-events

**Reply by markheath** [reply]
Ah I see, that could probably do with some clarification in the docs. Orchestrators are not kept alive in memory - they wake up, decide what to do next (based on the event sourced history) and go back to sleep. The situation its describing is when you send an orchestrator an "external" event *before* it's actually waiting for it. I *think* that event would survive the orchestrator sleeping and waking, although I haven't explicitly tested it. In any case, I don't think its a good pattern to send external events to orchestrators that aren't waiting for them. Usually you send them in response to something that happened earlier in the workflow, so your orchestrator can already be waiting before it gives anything the chance to send the external event.

**Reply by andsj073** [reply]
Agreed. In the case I am thinking of the orchestrator will be involved in an indefinitely (not necessarily long, but could be) long running workflow, which I guess is quite common. The event is the rendezvous of two processes obviously
